File: python/unique-binary-search-trees.py
# 96. 不同的二叉搜索树 - Medium
# https://leetcode-cn.com/problems/unique-binary-search-trees/
# 思路：
#   从1到n遍历每一个数字在根结点
#   F(i, n) = G(i - 1) * G(n - i)
#   长度为n的树以i为根结点的二叉搜索树个数 = 1 到 i - 1能构建的左子树的个数 * i + 1 到 n能构建的右子树的个数
#   
#   这个公式在数学上被称为卡塔兰数
#   C0 = 1, C(n + 1) = ((2*(2*n+1))/(n+2))C(n)
#

class Solution:
    def numTrees(self, n: int) -> int:
        C = 1
        for i in range(0, n):
            C = C * 2 * (2 * i + 1) / (i + 2)
        return int(C)

File: java/unique-binary-search-trees.java
// 96. 不同的二叉搜索树
// https://leetcode-cn.com/problems/unique-binary-search-trees/
package java;

class UniqueBinarySearchTrees {
    /**
     * 假设n个节点的二叉搜索树个数为 G(n), 以 i 为根节点的二叉搜索树个数为 f(i)
     * 则以i为根节点的左子树个数为 i - 1, 右子树个数为 n - i，
     * 所以 f(i) = G(i - 1) * G(n - i)
     * G(n) = G(0) * G(n - 1) + G(1) * G(n - 2) + ... + G(n - 1) * G(0)
     * 我们设 dp[i] 为 i 个节点时可形成的二叉搜索树的个数
     */
    public int numTrees(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            for (int j = 1; j < i + 1; j++) {
                dp[i] += dp[j - 1] * dp[i - j];
            }
        }
        return dp[n];
    }
}
